diagnostic_common_diagnostics/diagnostic_common_diagnostics/ntp_monitor.py

```diff
@@ -115,81 +77,135 @@ def __init__(self, ntp_hostname, offset=500, self_offset=500,
 
-    def pubCB(self, ev):
+    def pubCB(self):
         with self.mutex:
             if self.current_msg:
                 self.pub.publish(self.current_msg)
 
-    def checkCB(self, ev):
+    def checkCB(self):
         new_msg = DIAG.DiagnosticArray()
-        new_msg.header.stamp = rospy.get_rostime()
+        # new_msg.header.stamp = rospy.get_rostime()
 
-        st = ntp_diag(self.stat, self.ntp_hostname, self.offset, self.error_offset)
+        st = self.ntp_diag(self.stat, self.ntp_hostname, self.offset, self.error_offset)
         if st is not None:
             new_msg.status.append(st)
 
         if self.do_self_test:
-            st = ntp_diag(self.self_stat, self.hostname, self.self_offset, self.error_offset)
+            st = self.ntp_diag(self.self_stat, self.hostname, self.self_offset, self.error_offset)
             if st is not None:
                 new_msg.status.append(st)
 
         with self.mutex:
             self.current_msg = new_msg
 
-        self.checktimer = rospy.Timer(rospy.Duration(10), self.checkCB, True)
+
+
+    def ntp_diag(self, st, host, off, error_offset):
+        
+        def add_kv(stat_values, key, value):
+            kv = DIAG.KeyValue()
+            kv.key = key
+            kv.value = value
+            stat_values.append(kv)
+            
+        PROCESS_NAME = "ntpdate"
+        p = Popen([PROCESS_NAME, "-q", host], stdout=PIPE, stdin=PIPE, stderr=PIPE)
+        try:
+            stdout, stderr = p.communicate(timeout=5)
+        except TimeoutExpired:
+            p.kill()
+            stdout, stderr = p.communicate()
+            self.get_logger().error('Timeout running %s: "%s"' % (PROCESS_NAME, stderr))
+        rc = p.returncode
+        
+        st.values = []
+        add_kv(st.values, "Offset tolerance (us)", str(off))
+        add_kv(st.values, "Offset tolerance (us) for Error", str(error_offset))
+        output = stdout.decode("UTF-8")
+        errors = stderr.decode("UTF-8")
+        
+        if rc == 0:
+            re_match = re.search("offset (.*),", output)
+            measured_offset = float(re_match[1])*1000000
+
+            st.level = DIAG.DiagnosticStatus.OK
+            st.message = "OK"
+            add_kv(st.values, "Offset (us)", str(measured_offset))
+
+            if (abs(measured_offset) > off):
+                st.level = DIAG.DiagnosticStatus.WARN
+                st.message = "NTP Offset Too High"
+            if (abs(measured_offset) > error_offset):
+                st.level = DIAG.DiagnosticStatus.ERROR
+                st.message = "NTP Offset Too High"
+
+        else:
+            st.level = DIAG.DiagnosticStatus.ERROR
+            st.message = "Error Running ntpdate. Returned %d" % rc
+            add_kv(st.values, "Offset (us)", "N/A")
+            add_kv(st.values, "Output", output)
+            add_kv(st.values, "Errors", errors)
+
+        return st
+
 
 
 def ntp_monitor_main(argv=sys.argv):
```

Comment:
Okay, it was like this in the original code, but the `argv` here is basically ignored, because `args = parser.parse_args()` reads it again. I think it makes sense to pass argv to a main function but then also use it in [L186](https://github.com/ros/diagnostics/blob/be616105de061ae69222a1470a8d2ab953ad312b/diagnostic_common_diagnostics/diagnostic_common_diagnostics/ntp_monitor.py#L186)

Reply:
https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.parse_args

In python3, after trying to understand how exactly argparse works, I now understand the original intent of doing this; it's to override the default sys.argv. I'll add this back in.

Reply:
Done: 5ee2e11b48903e9621a37c0ff312d9cc7e25527a